google-cloud-firestore/src/main/java/com/google/cloud/firestore/encoding/RecordMapper.java

```diff
@@ -0,0 +1,237 @@
+ * annotations on record components. Since records are not supported in JDK versions < 16,
+ * reflection is used for inspecting record metadata.
+ */
+class RecordMapper<T> extends BeanMapper<T> {
+  private static final Logger LOGGER = Logger.getLogger(RecordMapper.class.getName());
+  private static final RecordInspector RECORD_INSPECTOR = new RecordInspector();
+
+  // Below are maps to find an accessor and constructor parameter index from a given property name.
+  // A property name is the name annotated by @PropertyName, if exists; or the component name.
+  // See method propertyName for details.
+  private final Map<String, Method> accessors = new HashMap<>();
+  private final Constructor<T> constructor;
+  private final Map<String, Integer> constructorParamIndexes = new HashMap<>();
+
+  RecordMapper(Class<T> clazz) {
+    super(clazz);
+
+    constructor = RECORD_INSPECTOR.getCanonicalConstructor(clazz);
+
+    AnnotatedElement[] recordComponents = RECORD_INSPECTOR.getRecordComponents(clazz);
+    if (recordComponents.length == 0) {
+      throw new RuntimeException("No properties to serialize found on class " + clazz.getName());
+    }
+
+    try {
+      for (int i = 0; i < recordComponents.length; i++) {
+        Field field = clazz.getDeclaredField(RECORD_INSPECTOR.getName(recordComponents[i]));
+        String propertyName = propertyName(field);
+        constructorParamIndexes.put(propertyName, i);
+        accessors.put(propertyName, RECORD_INSPECTOR.getAccessor(recordComponents[i]));
+        applyFieldAnnotations(field);
+      }
+    } catch (NoSuchFieldException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  Map<String, Object> serialize(T object, DeserializeContext.ErrorPath path) {
+    verifyValidType(object);
+    Map<String, Object> result = new HashMap<>();
+    for (Map.Entry<String, Method> entry : accessors.entrySet()) {
+      String property = entry.getKey();
+      // Skip @DocumentId annotated properties;
+      if (documentIdPropertyNames.contains(property)) {
+        continue;
+      }
+
+      Object propertyValue;
+      Method accessor = entry.getValue();
+      try {
+        propertyValue = accessor.invoke(object);
+      } catch (IllegalAccessException | InvocationTargetException e) {
+        throw new RuntimeException(e);
+      }
+
+      Object serializedValue = getSerializedValue(property, propertyValue, path);
+
+      result.put(property, serializedValue);
+    }
+    return result;
+  }
+
+  @Override
+  T deserialize(
+      Map<String, Object> values,
+      Map<TypeVariable<Class<T>>, Type> types,
+      DeserializeContext context) {
+    if (constructor == null) {
+      throw context.errorPath.deserializeError(
+          "Class "
+              + getClazz().getName()
+              + " does not define a no-argument constructor. If you are using ProGuard, make "
+              + "sure these constructors are not stripped");
+    }
```

Comment:
This check is redundant, since this class can't be instantiated with a null `constructor`. Also, it's not a no-argument constructor, but the record's canonical one.